save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/service/LokiLogService.kt

```diff
@@ -0,0 +1,98 @@
+package com.saveourtool.save.service
+
+import com.saveourtool.save.utils.StringList
+import com.saveourtool.save.utils.enableHuaweiProxy
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.fasterxml.jackson.databind.node.ObjectNode
+import org.springframework.web.reactive.function.client.WebClient
+import org.springframework.web.reactive.function.client.bodyToMono
+import reactor.core.publisher.Mono
+
+import java.time.Instant
+import java.time.LocalDateTime
+import java.time.ZoneOffset
+import java.time.format.DateTimeFormatterBuilder
+import java.time.temporal.ChronoField
+
+/**
+ * AgentLogService from Loki
+ */
+class LokiLogService(
+    lokiServiceUrl: String,
+    enableHuaweiProxy: Boolean = false,
+) : LogService {
+    private val webClient = WebClient.create(lokiServiceUrl)
+        .let {
+            if (enableHuaweiProxy) {
+                it.enableHuaweiProxy()
+            } else {
+                it
+            }
+        }
+
+    override fun get(containerName: String, from: Instant, to: Instant): Mono<StringList> = webClient.get()
+        .uri(
+            "/loki/api/v1/query_range?query={query}&start={start}&end={end}&direction=forward",
+            "{container_name=\"$containerName\"}",
+            from.toEpochNanoStr(),
+            to.toEpochNanoStr(),
+        )
+        .retrieve()
+        .bodyToMono<ObjectNode>()
+        .filter {
+            it["status"].asText() == "success"
+        }
+        .map { objectNode ->
+            objectNode["data"]
+                .elementsAsSequenceFrom("result")
+                .flatMap { it.elementsAsSequenceFrom("values") }
+                .map { jsonNode ->
+                    val (timestampNode, msgNode) = jsonNode.elementsAsSequence().toList()
+                    val timestamp = timestampNode.asText()
+                        .fromEpochNanoStr()
+                        .let {
+                            LocalDateTime.ofInstant(it, zoneId)
+                        }
+                    val msg = msgNode.asText()
+                    timestamp to msg
+                }
+                .sortedBy { (timestamp, _) -> timestamp }
```

Comment:
sortedBy is not required -- `direction=forward` controls it

Reply:
it's not clear, but according to doc result should be sorted, but lokicli sorts it again.
```
	if q.Forward {
		sort.Slice(allEntries, func(i, j int) bool { return allEntries[i].entry.Timestamp.Before(allEntries[j].entry.Timestamp) })
	} else {
		sort.Slice(allEntries, func(i, j int) bool { return allEntries[i].entry.Timestamp.After(allEntries[j].entry.Timestamp) })
	}
```

Reply:
left sorting as in loki/logcli